tensorflow/java/src/main/java/org/tensorflow/Operation.java

```diff
@@ -97,6 +99,46 @@ public Output output(int idx) {
     return new Output(this, idx);
   }
 
+  @Override
+  public int hashCode() {
+    Graph.Reference r = graph.ref();
+    try {
+      // xor the first and last 32 bits of the underlying 64-bit pointer value.
+      return (int) (unsafeNativeHandle ^ (unsafeNativeHandle >>> 32));
```

Comment:
One sequence is `hashset.add(stale_object)` -- it can succeed if `stale_object.hashCode()` happens to select an empty bucket, but cause a later `hashset.add(good_object)` to fail if `good_object.hashCode()` picks the same bucket and so triggers a `.equals()` comparison.

Pragmatically, raising exceptions on both hashCode() and equals() will generally fail quicker for stale objects within collections. But this is also an internal detail that can be changed if needed - happy to start the way you suggest and revisit if needed.

(The new object comment was just about the difference between Long.valueOf() and primitive longs, separate from whether or not we use Graph.ref(). I think here and in the other hashcode implementation trading-off this micro-optimization in favor of core hashcode calculations works for me as well.)